packages/playground/blueprints/src/lib/compile.ts

```diff
@@ -171,33 +195,92 @@ export function compileBlueprint(
 			networking: blueprint.features?.networking ?? false,
 		},
 		run: async (playground: UniversalPHP) => {
+			const stepStatus = new Map<StepDefinition, Promise<void>>();
+			const stepPromises = new Map<
+				StepDefinition,
+				{
+					resolve: (result: any) => void;
+					reject: (error: any) => void;
+				}
+			>();
+			for (const { step, failMode } of compiled) {
+				const promise = new Promise<void>((resolve, reject) => {
+					stepPromises.set(step, { resolve, reject });
+				});
+				stepStatus.set(step, promise);
+			}
+
 			try {
 				// Start resolving resources early
-				for (const { resources } of compiled) {
+				for (const { resources, step, failMode } of compiled) {
+					const stepPromise = stepPromises.get(step);
 					for (const resource of resources) {
 						resource.setPlayground(playground);
 						if (resource.isAsync) {
-							resource.resolve();
+							resource.resolve().catch(function (error) {
+								stepPromise?.reject(error);
+							});
 						}
 					}
 				}
 
-				for (const { run, step } of compiled) {
-					const result = await run(playground);
-					onStepCompleted(result, step);
+				/**
+				 * When a compiled step's `run` method is called, it
+				 * uses `resolveArguments` to await dependent resources.
+				 */
+				for (const { run, step, failMode } of compiled) {
+					const stepPromise = stepPromises.get(step);
+					let result;
+					try {
+						result = await run(playground);
+					} catch (error) {
+						stepPromise?.reject(result);
+						if (onStepError) {
+							onStepError(step, error);
+						} else if (failMode === 'abort') {
+							throw error;
+						}
+						/**
+						 * Skip - This allows the default shouldBoot() below to
+						 * consider the step completed.
+						 */
+						stepStatus.set(step, Promise.resolve());
+						continue;
+					}
+					stepPromise?.resolve(result);
+					onStepCompleted(step, result);
 				}
 			} finally {
-				try {
-					await (playground as any).goTo(
-						blueprint.landingPage || '/'
+				const completedStatus = new Map();
+				for (const { step } of compiled) {
+					completedStatus.set(
+						step,
+						await isSettledPromiseResolved(
```

Comment:
Will this await forever if this `finally` section is triggered by an error thrown before all the steps are run?